bin/wasm-node/javascript/src/index-browser.ts

```diff
@@ -121,23 +113,294 @@ function trustedBase64Decode(base64: string): Uint8Array {
+    // According to <https://w3c.github.io/webrtc-pc/#dom-rtcpeerconnection-generatecertificate>,
+    // browsers are guaranteed to support `{ name: "ECDSA", namedCurve: "P-256" }`.
+    RTCPeerConnection.generateCertificate({ name: "ECDSA", namedCurve: "P-256" } as EcKeyGenParams).then((localCertificate) => {
+      if (cancelOpening)
+        return;
+
+      // We need to build the multihash corresponding to the local certificate.
+      let localTlsCertificateMultihash: Uint8Array | null = null;
+      for (const { algorithm, value } of localCertificate.getFingerprints()) {
+        if (algorithm === 'sha-256') {
+          localTlsCertificateMultihash = new Uint8Array(34);
+          localTlsCertificateMultihash.set([0x12, 32], 0);
+          localTlsCertificateMultihash.set(value!.split(':').map((s) => parseInt(s, 16)), 2);
+          break;
+        }
+      }
+      if (localTlsCertificateMultihash === null) {
+        // Because we've already returned from the `connect` function at this point, we pretend
+        // that the connection has failed to open.
+        config.onConnectionClose('Failed to obtain the browser certificate fingerprint');
+        return;
+      }
+
+      // Create a new WebRTC connection.
+      pc = new RTCPeerConnection({ certificates: [localCertificate] });
+
+      // `onconnectionstatechange` is used to detect when the connection has closed or has failed
+      // to open.
+      // Note that smoldot will think that the connection is open even when it is still opening.
+      // Therefore we don't care about events concerning the fact that the connection is now fully
+      // open.
+      pc.onconnectionstatechange = (_event) => {
+        if (pc!.connectionState == "closed" || pc!.connectionState == "disconnected" || pc!.connectionState == "failed") {
+          config.onConnectionClose("WebRTC state transitioned to " + pc!.connectionState);
+
+          pc!.onconnectionstatechange = null;
+          pc!.onnegotiationneeded = null;
+          pc!.ondatachannel = null;
+
+          for (const channel of Array.from(dataChannels.values())) {
+            channel.onopen = null;
+            channel.onerror = null;
+            channel.onclose = null;
+            channel.onmessage = null;
+          }
+
+          pc!.close();  // Unclear whether this is necessary, but it doesn't hurt to do so.
+
+          for (const channel of Array.from(dataChannels.values())) {
+            channel.close();  // Unclear whether this is necessary, but it doesn't hurt to do so.
+          }
+          dataChannels.clear();
+        }
+      };
+
+      pc.onnegotiationneeded = async (_event) => {
+          // Create a new offer and set it as local description.
+          let sdpOffer = (await pc!.createOffer()).sdp!;
+          // The server excepts the ufrag and pwd to be the same. Modify the local description
+          // to ensure that.
+          // TODO: why does the server want that?
```

Comment:
should we link the spec here?

"_A_ generates a random string and sets that string as the username (_ufrag_
   or _username fragment_) and password on the SDP of the local offer."
   
 the answer to this question is: otherwise there's no way to learn `pwd` for B (server). Right now B receives ICE binding request, which contains ufrag (but no pwd). Since we've agreed that ufrag == pwd, B uses this to construct SDP locally.